jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/IOResources.java

```diff
@@ -398,11 +398,17 @@ protected Action process() throws Throwable
             BufferUtil.clearToFill(byteBuffer);
             if (remainingLength >= 0 && remainingLength < Integer.MAX_VALUE)
                 byteBuffer.limit((int)Math.min(byteBuffer.capacity(), remainingLength));
-            int read = channel.read(byteBuffer);
+            boolean eof = false;
+            while (byteBuffer.hasRemaining() && !eof)
+            {
+                int read = channel.read(byteBuffer);
+                if (read == -1)
+                    eof = true;
+                else if (remainingLength >= 0)
+                    remainingLength -= read;
+            }
             BufferUtil.flipToFlush(byteBuffer, 0);
-            if (remainingLength >= 0)
-                remainingLength -= byteBuffer.remaining();
-            terminated = read == -1 || remainingLength == 0;
+            terminated = eof || remainingLength == 0;
```

Comment:
Below in `onCompleteSuccess()` you can use `IO.close()` instead of the try/catch.
Probably the same for `onCompleteFailure()`, unless we really want to chain the exception (but it's just a close exception while handling another exception -- kinda who cares).

Reply:
Done.